README.md
---
Platform: OS X Yosemite ~ macOS Mojave

Language: Python | Bourne-Again Shell

Environment: Console | Terminal
---

&nbsp;

# MacDaily

[![Downloads](http://pepy.tech/badge/macdaily)](http://pepy.tech/count/macdaily)
[![version](https://img.shields.io/pypi/v/macdaily.svg)](https://pypi.org/project/macdaily)
[![format](https://img.shields.io/pypi/format/macdaily.svg)](https://pypi.org/project/macdaily)
[![status](https://img.shields.io/pypi/status/macdaily.svg)](https://pypi.org/project/macdaily)

[![language](https://img.shields.io/github/languages/top/JarryShaw/macdaily.svg)](https://github.com/JarryShaw/macdaily)
[![made-with-bash](https://img.shields.io/badge/Made%20with-Bash-1f425f.svg)](https://www.gnu.org/software/bash)
[![python](https://img.shields.io/pypi/pyversions/macdaily.svg)](https://python.org)
[![implementation](https://img.shields.io/pypi/implementation/macdaily.svg)](http://pypy.org)

&nbsp;

 - [About](#about)
 - [Installation](#install)
 - [Configuration](#configuration)
 - [Usage Manual](#usage)
    * [Start-Up](#startup)
    * [Commands](#command)
    * [Generals](#general)
 - [Troubleshooting](#issue)
 - [TODO](#todo)

---

&nbsp;

<a name="about"> </a>

## About

 > Package day-care manager on macOS.

&emsp; `macdaily` is a mediate collection of console scripts written in __Python__ and __Bourne-Again Shell__. Originally works as an automatic housekeeper for Mac to update all packages outdated, `macdaily` is now fully functioned and end-user oriented. Without being aware of everything about your Mac, one can easily work around and manage packages out of no pain using `macdaily`.

&nbsp;

<a name="install"> </a>

## Installation

&emsp; Just as many Python packages, `macdaily` can be installed through `pip` using the following command, which will get you the latest version from [PyPI](https://pypi.org).

```sh
pip install macdaily
```

&emsp; Or if you prefer the real-latest version and fetch from this Git repository, then the script below should be used.

```sh
git clone https://github.com/JarryShaw/macdaily.git
cd macdaily
pip install -e .
# and to update at any time
git pull
```

&emsp; And for tree format support in dependency command, you may need `pipdeptree`, then implicily you can use the following script to do so.

```sh
pip install macdaily[pipdeptree]
# or explicitly...
pip install macdaily pipdeptree
```

&emsp; Do please __NOTE__ that, `macdaily` runs only with support of Python from version ***3.5*** and on. And it shall only work ideally on ***macOS***.

&nbsp;

<a name="configuration"> </a>

## Configuration

 > This part might be kind of garrulous, for some may not know what's going on here. :wink:

&emsp; Since robust enough, `macdaily` now supports configuration upon user's own wish. One may set up log path, hard disk path, archive path and many other things, other than the default settings.

 > __NOTA BENE__ -- `macdaily` now supports configuration commands, see [Config Procedure](https://github.com/JarryShaw/MacDaily/tree/master/src#config) section for more information.

&emsp; The configuration file should lie under `~/.dailyrc`, which is hidden from Finder by macOS. To review or edit it, you may use text editors like `vim` and/or `nano`, or other graphic editors, such as `Sublime Text` and/or `Atom`, or whatever you find favourable.

```
[Path]
# In this section, paths for log files are specified.
# Please, under any circumstances, make sure they are valid.
logdir = ~/Library/Logs/MacDaily    ; path where logs will be stored
tmpdir = /tmp/dailylog              ; path where temporary runtime logs go
dskdir = /Volumes/Your Disk         ; path where your hard disk lies
arcdir = ${dskdir}/Developers       ; path where ancient logs archive

[Mode]
# In this section, flags for modes are configured.
# If you would like to disable the mode, set it to "false".
apm      = true     ; Atom packages
gem      = true     ; Ruby gems
mas      = true     ; Mac App Store applications
npm      = true     ; Node.js modules
pip      = true     ; Python packages
brew     = true     ; Homebrew Cellars
cask     = true     ; Caskroom Casks
dotapp   = true     ; Applications (*.app)
macapp   = true     ; all applications in /Application folder
system   = true     ; macOS system packages
cleanup  = true     ; cleanup caches
appstore = true     ; Mac App Store applications in /Application folder

[Daemon]
# In this section, scheduled tasks are set up.
# You may append and/or remove the time intervals.
update      = true      ; run update on schedule
uninstall   = false     ; don't run uninstall
reinstall   = false     ; don't run reinstall
postinstall = false     ; don't run postinstall
dependency  = false     ; don't run dependency
logging     = true      ; run logging on schedule
schedule    =           ; scheduled timing (in 24 hours)
    8:00                ; update & logging at 8:00
    22:30-update        ; update at 22:30
    23:00-logging       ; logging at 23:00

[Option]
# In this section, command options are picked.
# Do make sure these options are available for commands.
update  = --all --yes --pre --quiet --show-log --no-cask
logging = --all --quiet --show-log

[Account]
# In this section, account information are stored.
# You must not modify this part under any circumstances.
username = ...
password = ********

[Environment]
# In this section, environment specifications are set up.
# Please, under any circumstances, make sure all fields are valid.
bash-timeout = 1_000    ; timeout limit for each shell script in seconds
sudo-timeout = 5m       ; sudo command timeout as specified in /etc/sudoers
```

&emsp; Above is the default content of `.dailyrc`, following the grammar of `INI` files. Lines and words after number sign (`'#'`) and semicolon (`';'`) are comments, whose main purpose is to help understanding the contents of this file.

&emsp; In section `[Path]`, there are path names where logs and some other things to be stored. In section `[Mode]`, there are ten different modes to indicate if they are *enabled* or *disabled* when calling from `--all` option.

&emsp; You may wish to set the `dskdir` -- *path where your hard disk lies*, which allows `macdaily` to archive your ancient logs and caches into somewhere never bothers.

&emsp; Please __NOTE__ that, under all circumstances, of section `[Path]`, all values would better be a ***valid path name without blank characters*** (` \t\n\r\f\v`), except your hard disk `dskdir`.

&emsp; Besides, in section `[Daemon]`, you can decide which command is scheduled and when to run such command, with the format of `HH:MM[-CMD]`. The `CMD` is optional, which will be `any` if omits. And you may setup which command(s) will be registered as daemons and run with schedule through six booleans above. These boolean values help `macdaily` indicate which is to be launched when commands in `schedule` omit. That is to say, when `command` omits in `schedule`, `macdaily` will register all commands that set `true` in the above boolean values.

&emsp; Also, in section `[Option]`, you may set up optional arguments for the daemons above. Do please make sure these commands are **valid**. And if omit, an empty arguments will be given.

&emsp; Last but no least, in section `[Account]`, you should **NEVER** modify any contents under this section in order to keep `macdaily` working. However, you may setup this part with [`config`](https://github.com/JarryShaw/MacDaily/tree/master/src#config) command.

&nbsp;

<a name="usage"> </a>

## Usage Manual

<a name="startup"> </a>

### Start-Up

&emsp; Before we dive into the detailed usage of `macdaily`, let's firstly get our hands dirty with some simple commands.

 > __NOTE__ -- all acronyms and aliases are left out for a quick and clear view of `macdaily`

1. How to use `macdaily`?

    ```shell
    # call from $PATH
    $ macdaily [command ...] [flag ...]
    # or call from Python module
    $ python -m macdaily [command ...] [flag ...]
    ```

2. How to setup my disks and daemons?

    ```
    $ macdaily config
    ```

3. How to relaunch daemons after I manually modified `~/.dailyrc`?

    ```
    $ macdaily launch
    ```

4. How to archive ancient logs without running any commands?

    ```
    $ macdaily archive
    ```

5. How to update all outdated packages?

    ```
    $ macdaily update --all
    ```

6. How to update a certain package (eg: `hello` from Homebrew) ?

    ```
    $ macdaily update brew --package hello
    ```

7. How to uninstall a certain package along with its dependencies (eg: `pytest` from brewed CPython version 3.6) ?

    ```
    $ macdaily uninstall pip --brew --cpython --python_version=3 --package pytest
    ```

8. How to reinstall all packages but do not cleanup caches?

    ```
    $ macdaily reinstall --all --no-cleanup
    ```

9. How to postinstall packages whose name ranges between "start" and "stop" alphabetically?

    ```
    $ macdaily postinstall --all --startwith=start --endwith=stop
    ```

10. How to show dependency of a certain package as a tree (eg: `gnupg` from Homebrew) ?

    ```
    $ macdaily dependency brew --package gnupg --tree
    ```

11. How to log all applications on my Mac, a.k.a. `*.app` files?

    ```
    $ macdaily logging dotapp
    ```

12. How to run `macdaily` in quiet mode, i.e. with no output information (eg: `logging` in quiet mode) ?

    ```
    $ macdaily logging --all --quiet
    ```

13. How to dump a `Macfile` to keep track of all packages?

    ```
    $ macdaily bundle dump
    ```

<a name="command"> </a>

### Commands

&emsp; `macdaily` supports several different commands, from `archive`, `bundle`, `config`, `launch`, `update`, `unisntall`, `reinstall` and `postinstall` to `dependency` and `logging`. Of all commands, there are corresponding **aliases** for which to be reckoned as valid.


CHANGELOG.md
# Change Log

## [v2019.4.7.post1](https://github.com/JarryShaw/MacDaily/tree/v2019.4.7.post1) (2019-04-07)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.4.7.post1.c7250d-devel...v2019.4.7.post1)

**Merged pull requests:**

- New distribution \[2019.04.07.post1\] [\#57](https://github.com/JarryShaw/MacDaily/pull/57) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.4.7.post1.c7250d-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.4.7.post1.c7250d-devel) (2019-04-07)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.4.7...v2019.4.7.post1.c7250d-devel)

## [v2019.4.7](https://github.com/JarryShaw/MacDaily/tree/v2019.4.7) (2019-04-07)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.4.7.a0ca50-devel...v2019.4.7)

**Merged pull requests:**

-  New distribution \[2019.04.07\] [\#56](https://github.com/JarryShaw/MacDaily/pull/56) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.4.7.a0ca50-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.4.7.a0ca50-devel) (2019-04-07)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31.post1...v2019.4.7.a0ca50-devel)

## [v2019.3.31.post1](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31.post1) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31.post1.-devel...v2019.3.31.post1)

## [v2019.3.31.post1.-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31.post1.-devel) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31.post1.b8242a-devel...v2019.3.31.post1.-devel)

## [v2019.3.31.post1.b8242a-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31.post1.b8242a-devel) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.30.e89c90-devel...v2019.3.31.post1.b8242a-devel)

**Merged pull requests:**

-  New distribution \[2019.03.31.post1\] [\#55](https://github.com/JarryShaw/MacDaily/pull/55) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.3.30.e89c90-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.30.e89c90-devel) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.03.31...v2019.3.30.e89c90-devel)

## [v2019.03.31](https://github.com/JarryShaw/MacDaily/tree/v2019.03.31) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.03.30...v2019.03.31)

## [v2019.03.30](https://github.com/JarryShaw/MacDaily/tree/v2019.03.30) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.30...v2019.03.30)

## [v2019.3.30](https://github.com/JarryShaw/MacDaily/tree/v2019.3.30) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31.1b8496-devel...v2019.3.30)

**Merged pull requests:**

- New distribution \[2019.03.31\] [\#54](https://github.com/JarryShaw/MacDaily/pull/54) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.3.31.1b8496-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31.1b8496-devel) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31...v2019.3.31.1b8496-devel)

## [v2019.3.31](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31.e89c90-devel...v2019.3.31)

**Merged pull requests:**

- New distribution \[2019.03.30\] [\#53](https://github.com/JarryShaw/MacDaily/pull/53) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.3.31.e89c90-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31.e89c90-devel) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.31.-devel...v2019.3.31.e89c90-devel)

## [v2019.3.31.-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.31.-devel) (2019-03-30)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.28.post1...v2019.3.31.-devel)

## [v2019.3.28.post1](https://github.com/JarryShaw/MacDaily/tree/v2019.3.28.post1) (2019-03-28)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.28.post1.1baf18-devel...v2019.3.28.post1)

**Merged pull requests:**

- Bugfix [\#52](https://github.com/JarryShaw/MacDaily/pull/52) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.3.28.post1.1baf18-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.28.post1.1baf18-devel) (2019-03-28)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.28...v2019.3.28.post1.1baf18-devel)

## [v2019.3.28](https://github.com/JarryShaw/MacDaily/tree/v2019.3.28) (2019-03-28)
[Full Changelog](https://github.com/JarryShaw/MacDaily/compare/v2019.3.28.edfd7e-devel...v2019.3.28)

**Merged pull requests:**

- New distribution \[2019.03.28\] [\#51](https://github.com/JarryShaw/MacDaily/pull/51) ([JarryShaw](https://github.com/JarryShaw))

## [v2019.3.28.edfd7e-devel](https://github.com/JarryShaw/MacDaily/tree/v2019.3.28.edfd7e-devel) (2019-03-28)

src/update.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-


import argparse
import datetime
import os
import pathlib
import platform
import sys
import zipfile


from jsdaily.libupdate import *


# version string
__version__ = '0.10.4'


# today
today = datetime.datetime.today()


# display mode names
NAME = dict(
    apm = 'Atom',
    gem = 'Ruby',
    npm = 'Node.js',
    pip = 'Python',
    brew = 'Homebrew',
    cask = 'Caskroom',
    appstore = 'App Store',
)


# mode actions
MODE = dict(
    all = lambda *args, **kwargs: update_all(*args, **kwargs),
    apm = lambda *args, **kwargs: update_apm(*args, **kwargs),
    gem = lambda *args, **kwargs: update_gem(*args, **kwargs),
    npm = lambda *args, **kwargs: update_npm(*args, **kwargs),
    pip = lambda *args, **kwargs: update_pip(*args, **kwargs),
    brew = lambda *args, **kwargs: update_brew(*args, **kwargs),
    cask = lambda *args, **kwargs: update_cask(*args, **kwargs),
    cleanup = lambda *args, **kwargs: update_cleanup(*args, **kwargs),
    appstore = lambda *args, **kwargs: update_appstore(*args, **kwargs),
)


# terminal commands
python = sys.prefix             # Python version
program = ' '.join(sys.argv)    # arguments


# terminal display
red = 'tput setaf 1'    # blush / red
green = 'tput setaf 2'  # green
blue = 'tput setaf 14'  # blue
bold = 'tput bold'      # bold
under = 'tput smul'     # underline
reset = 'tput sgr0'     # reset


# error handling class
class UnsupoortedOS(RuntimeError):
    def __init__(self, message, *args, **kwargs):
        sys.tracebacklimit = 0
        super().__init__(message, *args, **kwargs)


def get_parser():
    parser = argparse.ArgumentParser(prog='jsupdate', description=(
                    'Automatic Package Update Manager'
                ), usage=(
                    'jsupdate [-hV] [-qv] [-fgm] [-a] [--[no-]MODE] MODE ... '
                ))
    parser.add_argument('-V', '--version', action='version', version=__version__)
    parser.add_argument('-a', '--all', action='append_const', const='all',
                        dest='mode', help=(
                            'update all packages installed through Atom, pip '
                            'RubyGem, Node.js, Homebrew, App Store, and etc'
                        ))

    parser.add_argument('--apm', action='append_const', const='apm', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--gem', action='append_const', const='gem', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--npm', action='append_const', const='npm', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--pip', action='append_const', const='pip', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--brew', action='append_const', const='brew', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--cask', action='append_const', const='cask', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--cleanup', action='append_const', const='cleanup', dest='mode', help=argparse.SUPPRESS)
    parser.add_argument('--appstore', action='append_const', const='appstore', dest='mode', help=argparse.SUPPRESS)

    parser.add_argument('--no-apm', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-gem', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-npm', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-pip', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-brew', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-cask', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-cleanup', action='store_true', default=False, help=argparse.SUPPRESS)
    parser.add_argument('--no-appstore', action='store_true', default=False, help=argparse.SUPPRESS)

    subparser = parser.add_subparsers(title='mode selection', metavar='MODE',
                        dest='mode', help=(
                            'update outdated packages installed through '
                            'a specified method, e.g.: apm, pip, brew, '
                            'cask, appstore, or alternatively and simply, '
                            'cleanup'
                        ))

    parser_apm = subparser.add_parser('apm', description=(
                            'Update Installed Atom Packages'
                        ), usage=(
                            'jsupdate apm [-h] [-qv] [-a] [-p PKG]'
                        ))
    parser_apm.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through apm'
                        ))
    parser_apm.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_apm.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser_apm.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))

    parser_gem = subparser.add_parser('gem', description=(
                            'Update Installed Ruby Packages'
                        ), usage=(
                            'jsupdate gem [-h] [-qv] [-a] [-p PKG]'
                        ))
    parser_gem.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through gem'
                        ))
    parser_gem.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_gem.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser_gem.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))

    parser_npm = subparser.add_parser('npm', description=(
                            'Update Installed Node.js Packages'
                        ), usage=(
                            'jsupdate npm [-h] [-qv] [-a] [-p PKG]'
                        ))
    parser_npm.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through npm'
                        ))
    parser_npm.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_npm.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser_npm.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))

    parser_pip = subparser.add_parser('pip', description=(
                            'Update Installed Python Packages'
                        ), usage=(
                            'jsupdate pip [-h] [-qv] [-bcsy] [-V VER] [-a] [-p PKG]'
                        ))
    parser_pip.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through pip'
                        ))
    parser_pip.add_argument('-V', '--python_version', action='store', metavar='VER',
                        choices=[
                            1, 2, 20, 21, 22, 23, 24, 25, 26, 27,
                            0, 3, 30, 31, 32, 33, 34, 35, 36, 37,
                        ], dest='version', type=int, default=0, help=(
                            'indicate which version of pip will be updated'
                        ))
    parser_pip.add_argument('-s', '--system', action='store_true', default=False,
                        dest='system', help=(
                            'update pip packages on system level, i.e. python '
                            'installed through official installer'
                        ))
    parser_pip.add_argument('-b', '--brew', action='store_true', default=False,
                        dest='brew', help=(
                            'update pip packages on Cellar level, i.e. python '
                            'installed through Homebrew'
                        ))
    parser_pip.add_argument('-c', '--cpython', action='store_true', default=False,
                        dest='cpython', help=(
                            'update pip packages on CPython environment'
                        ))
    parser_pip.add_argument('-y', '--pypy', action='store_true', default=False,
                        dest='pypy', help=(
                            'update pip packages on PyPy environment'
                        ))
    parser_pip.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_pip.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser_pip.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))

    parser_brew = subparser.add_parser('brew', description=(
                            'Update Installed Homebrew Packages'
                        ), usage=(
                            'jsupdate brew [-h] [-qv] [-fm] [-a] [-p PKG] [--no-cleanup]'
                        ))
    parser_brew.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through Homebrew'
                        ))
    parser_brew.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_brew.add_argument('-f', '--force', action='store_true', default=False,
                        help=(
                            'use "--force" when running `brew update`'
                        ))
    parser_brew.add_argument('-m', '--merge', action='store_true', default=False,
                        help=(
                            'use "--merge" when running `brew update`'
                        ))
    parser_brew.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser_brew.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))
    parser_brew.add_argument('--no-cleanup', action='store_false', default=True,
                        dest='nocleanup', help=(
                            'do not remove caches & downloads'
                        ))

    parser_cask = subparser.add_parser('cask', description=(
                            'Update Installed Caskroom Packages'
                        ), usage=(
                            'jsupdate cask [-h] [-qv] [-fg] [-a] [-p PKG] [--no-cleanup]'
                        ))
    parser_cask.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through Caskroom'
                        ))
    parser_cask.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_cask.add_argument('-f', '--force', action='store_true', default=False,
                        help=(
                            'use "--force" when running `brew cask upgrade`'
                        ))
    parser_cask.add_argument('-g', '--greedy', action='store_true', default=False,
                        help=(
                            'use "--greedy" when running `brew cask outdated`, '
                            'and directly run `brew cask upgrade --greedy`'
                        ))
    parser_cask.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser_cask.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))
    parser_cask.add_argument('--no-cleanup', action='store_false', default=True,
                        dest='nocleanup', help=(
                            'do not remove caches & downloads'
                        ))

    parser_cleanup = subparser.add_parser('cleanup', description=(
                            'Cleanup Caches & Downloads'
                        ), usage=(
                            'jsupdate cleanup [-h] [-q] [--no-brew] [--no-cask]'
                        ))
    parser_cleanup.add_argument('--no-gem', dest='gem', action='store_false', default=True,
                        help=(
                            'do not remove Ruby caches & downloads'
                        ))
    parser_cleanup.add_argument('--no-npm', dest='npm', action='store_false', default=True,
                        help=(
                            'do not remove Node.js caches & downloads'
                        ))
    parser_cleanup.add_argument('--no-pip', dest='pip', action='store_false', default=True,
                        help=(
                            'do not remove Python caches & downloads'
                        ))
    parser_cleanup.add_argument('--no-brew', dest='brew', action='store_false', default=True,
                        help=(
                            'do not remove Homebrew caches & downloads'
                        ))
    parser_cleanup.add_argument('--no-cask', dest='cask', action='store_false', default=True,
                        help=(
                            'do not remove Caskroom caches & downloads'
                        ))
    parser_cleanup.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))

    parser_appstore = subparser.add_parser('appstore', description=(
                            'Update installed App Store packages'
                        ), usage=(
                            'jsupdate appstore [-h] [-q] [-a] [-p PKG]'
                        ))
    parser_appstore.add_argument('-a', '--all', action='store_true', default=False,
                        dest='all', help=(
                            'update all packages installed through App Store'
                        ))
    parser_appstore.add_argument('-p', '--package', metavar='PKG', action='append',
                        dest='package', help=(
                            'name of packages to be updated, default is all'
                        ))
    parser_appstore.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))

    parser.add_argument('-f', '--force', action='store_true', default=False,
                        help=(
                            'run in force mode, only for Homebrew or Caskroom'
                        ))
    parser.add_argument('-m', '--merge', action='store_true', default=False,
                        help=(
                            'run in merge mode, only for Homebrew'
                        ))
    parser.add_argument('-g', '--greedy', action='store_true', default=False,
                        help=(
                            'run in greedy mode, only for Caskroom'
                        ))
    parser.add_argument('-q', '--quiet', action='store_true', default=False,
                        help=(
                            'run in quiet mode, with no output information'
                        ))
    parser.add_argument('-v', '--verbose', action='store_true', default=False,
                        help=(
                            'run in verbose mode, with detailed output information'
                        ))

    return parser


def main():
    if platform.system() != 'Darwin':
        raise UnsupoortedOS('update: script runs only on macOS')

    parser = get_parser()
    args = parser.parse_args()

    if args.mode is None:
        parser.print_help()
        return

    pathlib.Path('/tmp/log').mkdir(parents=True, exist_ok=True)
    pathlib.Path('/Library/Logs/Scripts/update').mkdir(parents=True, exist_ok=True)

    logdate = datetime.date.strftime(today, '%y%m%d')
    logname = f'/Library/Logs/Scripts/update/{logdate}.log'

    mode = '-*- Arguments -*-'.center(80, ' ')
    with open(logname, 'a') as logfile:
        logfile.write(datetime.date.strftime(today, '%+').center(80, '—'))
        logfile.write(f'\n\nCMD: {python} {program}')
        logfile.write(f'\n\n{mode}\n\n')
        for key, value in args.__dict__.items():
            logfile.write(f'ARG: {key} = {value}\n')

    if isinstance(args.mode, str):
        args.mode = [args.mode]
    for mode in args.mode:
        update = MODE.get(mode)
        log = update(args, file=logname, date=logdate)

    arcfile = '/Library/Logs/Scripts/archive.zip'
    filelist = list()
    with zipfile.ZipFile(arcfile, 'a', zipfile.ZIP_DEFLATED) as zf:
        abs_src = os.path.abspath('/Library/Logs/Scripts')
        for dirname, subdirs, files in os.walk('/Library/Logs/Scripts/update'):
            for filename in files:
                if filename == '.DS_Store':
                    continue
                name, ext = os.path.splitext(filename)
                if ext != '.log':
                    continue
                ctime = datetime.datetime.strptime(name, '%y%m%d')
                delta = today - ctime
                if delta > datetime.timedelta(7):
                    absname = os.path.abspath(os.path.join(dirname, filename))
                    arcname = absname[len(abs_src) + 1:]
                    zf.write(absname, arcname)
                    filelist.append(arcname)
                    os.remove(absname)

    mode = '-*- Update Logs -*-'.center(80, ' ')
    with open(logname, 'a') as logfile:
        logfile.write(f'\n\n{mode}\n\n')
        if not args.quiet:
            os.system(f'echo "-*- $({blue})Update Logs$({reset}) -*-"; echo ;')

        for mode in log:
            name = NAME.get(mode, mode)
            if log[mode] and all(log[mode]):
                pkgs = f', '.join(log[mode])
                logfile.write(f'LOG: Updated following {name} packages: {pkgs}.\n')
                if not args.quiet:
                    pkgs_coloured = f'$({reset}), $({red})'.join(log[mode])
                    os.system(f'echo "update: $({green}){mode}$({reset}): '
                              f'updated following $({bold}){name}$({reset}) packages: $({red}){pkgs_coloured}$({reset})"')
            else:
                logfile.write(f"LOG: No package updated in {name}.\n")
                if not args.quiet:
                    os.system(f'echo "update: $({green}){mode}$({reset}): '
                              f'no package updated in $({bold}){name}$({reset})"')

        if filelist:
            files = ', '.join(filelist)
            logfile.write(f'LOG: Archived following old logs: {files}\n')
            if not args.quiet:
                os.system(f'echo "update: $({green})cleanup$({reset}): '
                          f'ancient logs archived into $({under}){arcfile}$({reset})"')
        logfile.write('\n\n\n\n')


if __name__ == '__main__':
    sys.exit(main())
